File: server/datasets/controlled_benchmarks/java/multi_student_cohorts/cohort_c1/organic_7.java
/*
 * Graph BFS - Organic Submission #7
 * Runs BFS across every connected component so disconnected graphs are traversed.
 */

import java.util.*;

class Solution {
    public static List<String> bfsFrom(Map<String, List<String>> graph, String start, Set<String> visited) {
        Queue<String> queue = new ArrayDeque<>();
        visited.add(start);
        queue.add(start);
        List<String> component = new ArrayList<>();

        while (!queue.isEmpty()) {
            String node = queue.poll();
            component.add(node);
            List<String> neighbors = graph.getOrDefault(node, Collections.emptyList());
            for (String neighbor : neighbors) {
                if (!visited.contains(neighbor)) {
                    visited.add(neighbor);
                    queue.add(neighbor);
                }
            }
        }
        return component;
    }

    public static List<List<String>> bfsAllComponents(Map<String, List<String>> graph) {
        Set<String> visited = new HashSet<>();
        List<List<String>> components = new ArrayList<>();

        for (String node : graph.keySet()) {
            if (!visited.contains(node)) {
                components.add(bfsFrom(graph, node, visited));
            }
        }
        return components;
    }

    public static void main(String[] args) {
        Map<String, List<String>> graph = new HashMap<>();
        graph.put("A", Arrays.asList("B"));
        graph.put("B", Arrays.asList("A"));
        graph.put("C", Arrays.asList("D", "E"));
        graph.put("D", Arrays.asList("C"));
        graph.put("E", Arrays.asList("C"));
        graph.put("F", Collections.emptyList());

        System.out.println(bfsAllComponents(graph));
    }
}

File: server/datasets/controlled_benchmarks/python/multiple_files/multiple_c1/organic_7.py
"""
Graph BFS - Organic Submission #7
Runs BFS across every connected component so disconnected graphs
are fully traversed.
"""

from collections import deque


def bfs_from(graph, start, visited):
    queue = deque([start])
    visited.add(start)
    component = []

    while queue:
        node = queue.popleft()
        component.append(node)
        for neighbor in graph.get(node, []):
            if neighbor not in visited:
                visited.add(neighbor)
                queue.append(neighbor)

    return component


def bfs_all_components(graph):
    visited = set()
    components = []

    for node in graph:
        if node not in visited:
            components.append(bfs_from(graph, node, visited))

    return components


if __name__ == "__main__":
    graph = {
        "A": ["B"],
        "B": ["A"],
        "C": ["D", "E"],
        "D": ["C"],
        "E": ["C"],
        "F": [],
    }
    print(bfs_all_components(graph))
